HAL.md

```diff
@@ -0,0 +1,117 @@
+# Hardware abstract layer(HAL) overview
+
+## Abstract 
+Hardware abstract layer(HAL) defines a standard interface for hardware vendors to implement, which enables Tangle-Accelerator to be aware of lower-level driver implementations. HAL layer support such as uart, GPIO, secure storage and other low-level operators used in TA-endpoint. The TA-endpoint is a daemon which base on Linux which offer monitor for device and can send the message to tangle-accelerator. HAL layer lets user only needs to implement their device operators from abstract interface which defined inside `device.h` but not need to know how TA-endpoint work. 
+
+## Directory Overview
+* connectivity: Connect to Tangle-accelerator functions.
+* devices: The real directory to implement devices.
+* hal: Hardware abstraction layer.
+* hooks: Git hook functions.
+* tests: Unit-tests.
+* third_party: Third party directory.
+* utils: Utility functions for TA-endpoint.
+
+## Create directory for new device
+Create directory for new device into `devices`.
+```
+mkdir devices/<my_device>
+```
+* Create `impl.c` and `Makefile` under new device directory.
+* Include `device.h` headers into new device header file.
+
+Implement new device operations, which are included from `hal/device.h`
+
+## Quick view inside device.h
+### Abstraction 
+Here are some operations needed to implement inside new device:
+* device_operations
+    * init : initialize device
+    * fini : finalize device
+    * get_key : get device key
+    * get_device_id : get device id(IMEI or other identifier)
+* uart_operations
+    * init : initialize uart
+    * write : write command to uart device
+    * read : read from uart device 
+    * clean : flush buffer
+* secure_store_operations
+    * init : initialize secure storage
+    * write : write item to secure storage
+    * read : read item from secure storage
+    * delete : delete item inside secure storage
+### Functions to use inside `impl.c`
+Here are some functions needed to use inside `impl.c`:
+* register_device : register device on startup
+* unregister_device : unregister device
+* DECLARE_DEVICE : this must been declare inside `impl.c` 
+
+### Modify hal Makefile
+Add new device into `hal/Makefile`:
+
+* Add device object after DEVICES_OBJ
+* Add new device build target
+```=
+DEVICES_OBJ = wp7702.o emulator.o device.o mydevice.o
+INCLUDES += -I/$(CURDIR)
+export DEVICES_OBJ
+
+all: $(DEVICES_OBJ)
+... 
+mydevice.o: device.o
+    $(MAKE) -C ../devices/mydevice
+...
+    
+```
+
+### Implement new device
+Implement new device which create under devices directory, edit Makefile, example device name as mydevice:
```

Comment:
Implement new device which `is created` under the devices directory `and edit the` Makefile.
`The example device is named` as mydevice: